File: toy_ray_tracer/src/scenes/cornell_box_foggy.rs
use crate::camera::{Camera, CameraOpt};
use crate::hittable::Hittable;
use crate::hittable_list::HittableList;
use crate::materials::{DiffuseLight, Lambertian};
use crate::scene::{RenderOptions, Scene};
use crate::shapes::{AARect, ConstantMedium, Cube, Plane};
use crate::textures::ConstantTexture;
use crate::transforms::Translate;
use crate::transforms::{Axis, Rotate};
use crate::vec::{vec3, Vec3};

fn create_world() -> Box<dyn Hittable> {
    let red = Lambertian::new(ConstantTexture::new(0.65, 0.05, 0.05));
    let white = Lambertian::new(ConstantTexture::new(0.73, 0.73, 0.73));
    let green = Lambertian::new(ConstantTexture::new(0.12, 0.45, 0.15));
    let light = DiffuseLight::new(ConstantTexture::new(7.0, 7.0, 7.0));
    let mut world = HittableList::new();
    world.add(AARect::new(Plane::YZ, 0.0, 555.0, 0.0, 555.0, 555.0, green));
    world.add(AARect::new(Plane::YZ, 0.0, 555.0, 0.0, 555.0, 0.0, red));
    world.add(AARect::new(
        Plane::ZX,
        127.0,
        432.0,
        113.0,
        443.0,
        554.0,
        light,
    ));
    world.add(AARect::new(
        Plane::ZX,
        0.0,
        555.0,
        0.0,
        555.0,
        555.0,
        white.clone(),
    ));
    world.add(AARect::new(
        Plane::ZX,
        0.0,
        555.0,
        0.0,
        555.0,
        0.0,
        white.clone(),
    ));
    world.add(AARect::new(
        Plane::XY,
        0.0,
        555.0,
        0.0,
        555.0,
        555.0,
        white.clone(),
    ));
    let box1 = Translate::new(
        Rotate::new(
            Axis::Y,
            Cube::new(
                Vec3::new(0.0, 0.0, 0.0),
                Vec3::new(165.0, 165.0, 165.0),
                white.clone(),
            ),
            -18.0,
        ),
        Vec3::new(130.0, 0.0, 65.0),
    );
    let box2 = Translate::new(
        Rotate::new(
            Axis::Y,
            Cube::new(
                Vec3::new(0.0, 0.0, 0.0),
                Vec3::new(165.0, 330.0, 165.0),
                white,
            ),
            15.0,
        ),
        Vec3::new(265.0, 0.0, 295.0),
    );
    world.add(ConstantMedium::new(
        box1,
        0.01,
        ConstantTexture::new(1.0, 1.0, 1.0),
    ));
    world.add(ConstantMedium::new(
        box2,
        0.01,
        ConstantTexture::new(0.0, 0.0, 0.0),
    ));
    Box::new(world)
}

pub fn create_scene(opt: RenderOptions) -> Scene {
    let camera = Camera::new(CameraOpt {
        look_from: Vec3::new(278.0, 278.0, -800.0),
        look_at: Vec3::new(278.0, 278.0, 0.0),
        view_up: vec3::YUP,
        vertical_fov: 40.0,
        aspect: opt.aspect(),
        aperture: 0.0,
        focus_dist: 10.0,
        time0: 0.0,
        time1: 1.0,
    });

    let world = create_world();

    return Scene::new(
        camera,
        world,
        Vec3::new(0.7, 0.8, 1.0),
        String::from("cornell_box_foggy"),
        String::from("cornell box with fogs"),
    );
}

File: assets/projects/cornell_box_foggy.js
export function create_world() {
  const red = make_material({
    kind: "lambertian",
    albedo: {
      kind: "constant_texture",
      color: [0.65, 0.05, 0.05],
    },
  });

  const white = make_material({
    kind: "lambertian",
    albedo: {
      kind: "constant_texture",
      color: [0.73, 0.73, 0.73],
    },
  });

  const green = make_material({
    kind: "lambertian",
    albedo: {
      kind: "constant_texture",
      color: [0.12, 0.45, 0.15],
    },
  });

  const light = make_material({
    kind: "diffuse_light",
    emit: {
      kind: "constant_texture",
      color: [7.0, 7.0, 7.0],
    },
  });

  const world = make_geometry_list([]);

  world.push({
    kind: "rect",
    v0: [555, 0, 0],
    v1: [555, 555, 555],
    material: green,
  });

  world.push({
    kind: "rect",
    v0: [0, 0, 0],
    v1: [0, 555, 555],
    material: red,
  });

  world.push({
    kind: "flip_face",
    child: {
      kind: "rect",
      v0: [113, 554, 127],
      v1: [443, 554, 432],
      // radius: 100,
      material: light,
    },
  });

  world.push({
    kind: "rect",
    v0: [0, 555, 0],
    v1: [555, 555, 555],
    material: white,
  });

  world.push({
    kind: "rect",
    v0: [0, 0, 0],
    v1: [555, 0, 555],
    material: white,
  });

  world.push({
    kind: "rect",
    v0: [0, 0, 555],
    v1: [555, 555, 555],
    material: white,
  });

  // cubes
  let cube1 = make_geometry({
    kind: "translate",
    offset: [130, 0, 65],
    child: {
      kind: "rotate",
      axis: "Y",
      angle: -18.0,
      child: {
        kind: "cube",
        p_min: [0, 0, 0],
        p_max: [165, 165, 165],
        material: white,
      },
    },
  });

  let cube2 = make_geometry({
    kind: "translate",
    offset: [265, 0, 295],
    child: {
      kind: "rotate",
      axis: "Y",
      angle: 15.0,
      child: {
        kind: "cube",
        p_min: [0, 0, 0],
        p_max: [165, 330, 165],
        material: white,
      },
    },
  });

  cube1 = make_geometry({
    kind: "constant_medium",
    boundary: cube1,
    density: 0.01,
    texture: {
      kind: "constant_texture",
      color: [1.0, 1.0, 1.0],
    },
  });

  cube2 = make_geometry({
    kind: "constant_medium",
    boundary: cube2,
    density: 0.01,
    texture: {
      kind: "constant_texture",
      color: [0.0, 0.0, 0.0],
    },
  });

  world.push(cube1, cube2);

  return world;
}

export default make_project({
  name: "cornell_box_foggy",
  settings: {
    output_dir: "./output",
    height: 800,
    width: 800,
    nsamples: 100,
    max_depth: 15,
  },
  scene: {
    camera: {
      look_from: [278, 278, -800],
      look_at: [278, 278, 0],
      view_up: [0, 1, 0],
      vertical_fov: 40.0,
      aspect: 1.0,
      aperture: 0.0,
      focus_dist: 10.0,
      time0: 0.0,
      time1: 0.0,
    },
    sky: {
      kind: "solid",
      //   background: [0.7, 0.8, 1.0],
      background: [0.0, 0.0, 0.0],
    },
    world: {
      // kind: "list",
      kind: "bvh",
      children: create_world(),
      time0: 0,
      time1: 1.0,
    },
  },
});
